transport/x/cherami/inbound.go

```diff
@@ -0,0 +1,199 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cherami
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/errors"
+	"go.uber.org/yarpc/internal/sync"
+	"go.uber.org/yarpc/serialize"
+	"go.uber.org/yarpc/transport/x/cherami/internal"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/uber/cherami-client-go/client/cherami"
+)
+
+const (
+	transportName = "cherami"
+
+	defaultPrefetchCount = 10
+
+	defaultCheramiTimeout = 15 * time.Second
+)
+
+// InboundConfig defines the config in order to create a Inbound.
+//
+// PrefetchCount controls the number of messages to buffer locally. Inbounds
+// which process messages very fast may want to specify larger value for
+// PrefetchCount for faster throughput. On the flip side larger values for
+// PrefetchCount will result in more messages being buffered locally causing
+// high memory footprint.
+type InboundConfig struct {
+	Destination   string
+	ConsumerGroup string
+	PrefetchCount int
+	Timeout       time.Duration
+}
+
+// Inbound receives Oneway YARPC requests over Cherami.
+type Inbound struct {
+	config        InboundConfig
+	consumer      cherami.Consumer
+	router        transport.Router
+	tracer        opentracing.Tracer
+	client        cherami.Client
+	clientFactory internal.ClientFactory
+
+	once sync.LifecycleOnce
+}
+
+// NewInbound builds a new Cherami inbound.
+func (t *Transport) NewInbound(config InboundConfig) *Inbound {
+	if config.PrefetchCount == 0 {
+		config.PrefetchCount = defaultPrefetchCount
+	}
+	if config.Timeout/time.Second <= 0 {
+		config.Timeout = defaultCheramiTimeout
+	}
+	return &Inbound{
+		config:        config,
+		tracer:        t.tracer,
+		client:        t.client,
+		clientFactory: t.clientFactory,
+	}
+}
+
+// Transports returns nil for now
+func (i *Inbound) Transports() []transport.Transport {
+	return nil
+}
+
+// SetRouter configures a router to handle incoming requests.
+// This satisfies the transport.Inbound interface, and would be called
+// by a dispatcher when it starts.
+func (i *Inbound) SetRouter(router transport.Router) {
+	i.router = router
+}
+
+// IsRunning returns whether the inbound is still processing requests.
+func (i *Inbound) IsRunning() bool {
+	return i.once.IsRunning()
+}
+
+// Start starts the inbound, reading and handling messages from cherami
```

Comment:
general doc nits:

- Cherami is a proper noun in this context. You should capitalize it.
- Docs should use full sentences; they should end with a period.